### PR TITLE
Fix/improve parts of the Tobira harvest API (subtitles & duration)

### PR DESCRIPTION
**Pull request by LukasKalbertodt**
Fixes https://github.com/elan-ev/tobira/issues/960
Fixes https://github.com/elan-ev/tobira/issues/884

Look at the commits for more information. 

Note: I was not able to test with events that actually contain subtitles! Since @KatrinIhler asked for this (for Bern), it would be very very cool if you could test this. In particular, whether it fixes the issue you had.

And if @Arnei has time to take a look at this, I'd be interested in whether you see any problems with my "find subtitles" logic (second commit).

### Your pull request should…

* [ ] have a concise title
* [ ] [close an accompanying issue](https://help.github.com/en/articles/closing-issues-using-keywords) if one exists
* [ ] [be against the correct branch](https://docs.opencast.org/develop/developer/development-process#acceptance-criteria-for-patches-in-different-versions)
* [ ] include migration scripts and documentation, if appropriate
* [ ] pass automated tests
* [ ] have a clean commit history
* [ ] [have proper commit messages (title and body) for all commits](https://medium.com/@steveamaza/e028865e5791)

### CODE REVIEW

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/Item.java [hunk 1]
```diff
@@ -264,41 +275,35 @@ private static List<Jsons.Val> findCaptions(MediaPackage mp) {
...
-              captions.get(uri).add(suffix);
+        .map(track -> {
+          final var tags = track.getTags();
+          final Function<String, Optional<String>> findTag = (String prefix) -> Arrays.stream(tags)
+                .filter(tag -> tag.startsWith(prefix))
+                .map(tag -> tag.substring(prefix.length()))
+                .findFirst();
+
+          // Try to get a language for this subtitle track. We first check the proper tag.
+          var lang = findTag.apply("type");
```
Comment (Arnei) [inline modules/tobira/src/main/java/org/opencastproject/tobira/impl/Item.java]: Probably want to check for `lang` here, not `type`

Comment (LukasKalbertodt) [reply modules/tobira/src/main/java/org/opencastproject/tobira/impl/Item.java]: Wuppsi! That's indeed a stupid copy&paste mistake :D Thanks for catching!

**Comment by KatrinIhler** [comment]
Is this still compatible with the "old" way of handling captions, namely as attachments? Bc Arne's subtitle changes are only in OC 15+ afaik (and our Frankenstein OC 13 :sweat_smile: ).

**Comment by LukasKalbertodt** [comment]
Yep, it is still compatible. See the commit message of the relevant (second) commit.

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/Item.java [hunk 2]
```diff
@@ -264,41 +275,35 @@ private static List<Jsons.Val> findCaptions(MediaPackage mp) {
...
-          // Try to get language from flavor. Otherwise there is an empty list
-          // in the map.
-          final var subflavor = track.getFlavor().getSubtype();
-          if (subflavor.startsWith("vtt+")) {
-            final var suffix = subflavor.substring("vtt+".length());
-            if (suffix.length() > 0) {
-              captions.get(uri).add(suffix);
+        .map(track -> {
+          final var tags = track.getTags();
+          final Function<String, Optional<String>> findTag = (String prefix) -> Arrays.stream(tags)
```
Comment (KatrinIhler) [inline modules/tobira/src/main/java/org/opencastproject/tobira/impl/Item.java:280]: This doesn't actually work bc `generator` and `generator-type` have the same beginning and this also leaves in the ":".

```suggestion
          final Function<String, Optional<String>> findTag = (String prefix) -> Arrays.stream(tags)
                .map(tag -> tag.split(":"))
                .filter(tagArray -> (tagArray.length == 2 && tagArray[0].equals(prefix)))
                .map(tagArray -> tagArray[1])
                .findFirst();
```

File: modules/tobira/src/main/java/org/opencastproject/tobira/impl/Item.java [hunk 3]
```diff
@@ -264,41 +275,35 @@ private static List<Jsons.Val> findCaptions(MediaPackage mp) {
...
+                .filter(tag -> tag.startsWith(prefix))
+                .map(tag -> tag.substring(prefix.length()))
+                .findFirst();
```
Comment (KatrinIhler) [inline modules/tobira/src/main/java/org/opencastproject/tobira/impl/Item.java]: This doesn't actually work bc `generator` and `generator-type` have the same beginning and this also leaves in the ":".

```suggestion
          final Function<String, Optional<String>> findTag = (String prefix) -> Arrays.stream(tags)
                .map(tag -> tag.split(":"))
                .filter(tagArray -> (tagArray.length == 2 && tagArray[0].equals(prefix)))
                .map(tagArray -> tagArray[1])
                .findFirst();
```

Comment (LukasKalbertodt) [reply modules/tobira/src/main/java/org/opencastproject/tobira/impl/Item.java]: Woops :D Should be fixed

**Comment by LukasKalbertodt** [comment]
Spurious CI failure. If someone can restart that, that'd be great.

**Comment by KatrinIhler** [comment]
> Spurious CI failure. If someone can restart that, that'd be great.

I have the power :magic_wand:

**Review by KatrinIhler — APPROVED** [review]
Code looks good and is working correctly on our test system (with captions).

**Pull request merged**